src/stan/io/random_var_context.hpp

```diff
@@ -0,0 +1,221 @@
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace io {
+
+    /**
+     * This is an implementation of a var_context that initializes
+     * the unconstrained values randomly. This is used for initialization.
+     */
+    class random_var_context : public var_context {
+    public:
+      /**
+       * Destructor.
+       *
+       */
+      ~random_var_context() {}
+
+      /**
+       * Constructor.
+       *
+       * On construction, this var_context will generate random
+       * numbers on the unconstrained scale for the model provided.
+       * Once generated, the class is immutable.
+       *
+       * This class only generates values for the parameters in the
+       * Stan program and does not generate values for transformed parameters
+       * or generated quantities.
+       *
+       * @tparam Model a generated Stan program or a class that follows
+       *   that template
+       * @tparam RNG Random number generator type. Typically
+       *   boost::ecuyer1988
+       * @param model instantiated model to generate variables for
+       * @param[in,out] rng pseudo-random number generator
+       * @param[in] init_radius the unconstrained variables are uniform draws
+       *   from -init_radius to init_radius. If this is 0, the unconstrained
+       *   values are set to 0. If the init_radius is non-zero, it is expected
+       *   to be greater than 0.
+       */
+      template <class Model, class RNG>
+      random_var_context(Model& model,
+                         RNG& rng,
+                         double init_radius,
+                         bool init_zero)
+        : unconstrained_params_(model.num_params_r()) {
+        size_t num_unconstrained_ = model.num_params_r();
+        model.get_param_names(names_);
+        model.get_dims(dims_);
+
+        // cutting names_ and dims_ down to just the constrained parameters
+        std::vector<std::string> constrained_params_names;
+        model.constrained_param_names(constrained_params_names, false, false);
+        size_t keep = constrained_params_names.size();
+        size_t index = 0;
+        size_t num = 0;
+        for (index = 0; index < dims_.size(); index++) {
+          size_t size = 1;
+          for (size_t n = 0; n < dims_[index].size(); n++)
+            size *= dims_[index][n];
+          num += size;
+          if (num > keep)
+            break;
+        }
+        dims_.erase(dims_.begin() + index, dims_.end());
+        names_.erase(names_.begin() + index, names_.end());
+
+        if (init_zero) {
+          for (size_t n = 0; n < num_unconstrained_; n++)
+            unconstrained_params_[n] = 0.0;
+        } else {
+          boost::random::uniform_real_distribution<double>
+            unif(-init_radius, init_radius);
+          for (size_t n = 0; n < num_unconstrained_; n++)
+            unconstrained_params_[n] = unif(rng);
+        }
+
+        std::vector<double> constrained_params;
+        std::vector<int> int_params;
```

Comment:
The model code still has a vector of integer parameters
awaiting some implementation.  I should've never coded that
in the first place ahead of any plan to implement it.
But they are params in the usual Stan sense (as opposed to
arguments to probabilty functions, which are always called
"parameters" even if they're data).

> On Aug 9, 2016, at 7:53 AM, Krzysztof Sakrejda <notifications@github.com> wrote:
> 
> In src/stan/io/random_var_context.hpp:
> 
> > +        }
> > +        dims_.erase(dims_.begin() + index, dims_.end());
> > +        names_.erase(names_.begin() + index, names_.end());
> > +
> > +        if (init_zero) {
> > +          for (size_t n = 0; n < num_unconstrained_; n++)
> > +            unconstrained_params_[n] = 0.0;
> > +        } else {
> > +          boost::random::uniform_real_distribution<double>
> > +            unif(-init_radius, init_radius);
> > +          for (size_t n = 0; n < num_unconstrained_; n++)
> > +            unconstrained_params_[n] = unif(rng);
> > +        }
> > +
> > +        std::vector<double> constrained_params;
> > +        std::vector<int> int_params;
> 
> I guess this is "params" in some non-Stan sense since we don't have integer parameters? Just confusing every time I see it.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>